crates/services/sync/src/import.rs

```diff
@@ -241,15 +214,110 @@ where
         // Count the number of successfully executed blocks and
         // find any errors.
         // Fold the stream into a count and any errors.
-        .fold((0usize, Ok(())), |(count, err), result| async move {
+        .fold((0usize, Ok(())), |(count, res), result| async move {
             match result {
-                Ok(_) => (count + 1, err),
+                Ok(_) => (count + 1, res),
                 Err(e) => (count, Err(e)),
             }
         })
         .in_current_span()
         .await
     }
+
+    async fn get_block_for_header(
+        result: anyhow::Result<SourcePeer<SealedBlockHeader>>,
+        p2p: Arc<P>,
+        consensus_port: Arc<C>,
+    ) -> anyhow::Result<Option<SealedBlock>> {
+        let header = match result {
+            Ok(h) => h,
+            Err(e) => return Err(e),
+        };
+        let SourcePeer {
+            peer_id,
+            data: header,
+        } = header;
+        let id = header.entity.id();
+        let block_id = SourcePeer { peer_id, data: id };
+
+        // Check the consensus is valid on this header.
+        if !consensus_port
+            .check_sealed_header(&header)
+            .trace_err("Failed to check consensus on header")?
+        {
+            tracing::warn!("Header {:?} failed consensus check", header);
+            return Ok(None)
+        }
+
+        // Wait for the da to be at least the da height on the header.
+        consensus_port
+            .await_da_height(&header.entity.da_height)
+            .await?;
+
+        get_transactions_on_block(p2p.as_ref(), block_id, header).await
+    }
+}
+
+fn get_header_range<P: PeerToPeerPort + Send + Sync + 'static>(
+    range: RangeInclusive<u32>,
+    p2p: Arc<P>,
+    batch_size: u32,
+    requests_in_flight: usize,
+) -> impl Stream<Item = anyhow::Result<SourcePeer<SealedBlockHeader>>> {
+    futures::stream::iter(range_chunks(range, batch_size))
+        .map(move |range| {
+            tracing::debug!(
+                "getting header range from {} to {} inclusive",
+                range.start(),
+                range.end()
+            );
+            let p2p = p2p.clone();
+            async move {
+                let headers = get_header_range_single_req(&range, p2p)
+                    .await
+                    .trace_err("Failed to get headers")?;
+                Ok((range, headers))
+            }
+            .instrument(tracing::debug_span!("get_header_range_single_req"))
+            .in_current_span()
+        })
+        .buffered(requests_in_flight)
+        .into_scan_err()
+        .scan_err()
+        .map(|val| {
+            let sorted_headers = match val {
+                Ok((mut range, headers)) => headers
+                    .into_iter()
+                    .map(move |header| {
+                        let height = range.next().unwrap();
```

Comment:
If the `.unwrap()` is guaranteed to be safe, can we use `expect(..)` with a message explaining this instead?

Reply:
Oh. Yikes. Good catch. I think I'll just `.and_then` the following logic. That's good anyway because if we have a response that is longer than our `range` all the extra headers should be ignored`.

Reply:
we can make a todo to retry batches at some point, but if this becomes an issue we can just reduce the max number of buffered header batches to avoid too much throw-away work.